Infinite slider

lukaszfranczak opened the pull request:
The infinite slider is finally functional - however a misbehaviour is present as when Prev button is clicked there is a delay in photo switch as "right" attribute is changed twice. This only happens when we click on a Prev button immediately after opening the website - in other conditions it (seems to) works fine.

lukaszfranczak committed:
Modul 04 - jQuery - progress bar

lukaszfranczak committed:
Progress Bar - added animation

lukaszfranczak committed:
Progress Bar - conflict solve - added animation

lukaszfranczak committed:
Modul 04 - infinite slider - fail

lukaszfranczak committed:
Modul 04 - progress bar - code reviewed

lukaszfranczak committed:
Modul 04 - jQuery - form, parallax

lukaszfranczak committed:
Modul 04 - infinite slider - functional

tomecko commented on 4_Zadania_dodatkowe/4.ProgressBar/js/app.js:
```diff
@@ -1,5 +1,5 @@
 $(document).ready(function () {
-   // console.log('test');
+    // console.log('test');
```
why not remove? this line and other `console.log`s or redundant comments

lukaszfranczak replied:
all redundant comments removed

tomecko commented on 4_Zadania_dodatkowe/4.ProgressBar/js/app.js:
```diff
@@ -14,17 +14,12 @@ $(document).ready(function () {
-        //     }
-        // });
+        barProgress
+            .filter('#'+clickedBarId)
```
so actually this can be simplified a little bit
there is no need in creating `barProgress` variable
you can remove line 7 and write `$('#' + clickedBarId)` (as there should be no two elements with the same `id`)

lukaszfranczak replied:
thanks! I checked other scripts as well - in all cases where a variable is used only once I removed them

tomecko commented on 4_Zadania_dodatkowe/4.ProgressBar/js/app.js:
```diff
@@ -14,17 +14,12 @@ $(document).ready(function () {
+            .filter('#'+clickedBarId)
+            .removeClass(['blue', 'orange', 'green'])
+            .addClass(clickedBarColor)
+            .animate({width: clickedBarWidth});
```
one may say that creating all these `clickedBar...` variables is not needed as well
they are used only once so you could also write `.addClass($(this).data('color'))` and so on

lukaszfranczak replied:
same as above

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/css/style.css:
```diff
@@ -1,10 +1,21 @@
+.slider ul{
+  position: relative;
   list-style: none;
+  width: 3228px;
```
remove

lukaszfranczak replied:
Hm. When line 10 is removed nothing is presented on the website so it's not working as it should be

tomecko replied:
I'd say you need to set it in js.

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/css/style.css:
```diff
@@ -1,10 +1,21 @@
-.slider {
-  height: 250px;
+.slider{
+  height: 200px; /* zmiana z 250 na 200 bo przy 250 widac czesc 2 obrazka */
```
:+1:

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/css/style.css:
```diff
@@ -1,10 +1,21 @@
+  width: 3228px;
+  margin: 0;
+  padding: 0;
+  right: 404px;
```
I think it should be handled in `js`

lukaszfranczak replied:
updated - now both right and width is set by jquery

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,71 @@
+    // console.log(elementUl);
+
+    var visibleImageIndex = 1;
+    var imageWidth = 404;
```
your script should not assume image's width
you can read it from matched element using http://api.jquery.com/outerWidth/
this is because your script should work well with images of different size without modification

lukaszfranczak replied:
I tried outerwidth(), css('width') combined with reset as well but nonetheless I still get 400px instead of 404px so the effect is not satisfactory (now I get the width and add 4px manually - still not how it should be done)

tomecko replied:
I think these extra 4 pixels are caused by a whitespace between `<li>` elements.
See https://css-tricks.com/fighting-the-space-between-inline-block-elements/ for explanation and possible solutions. I'd say you may want to use `font-size` technique in this case.

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,71 @@
+    // console.log(firstElement);
+    // console.log(lastElement);
+    elementUl.append(firstElement);
+    elementUl.prepend(lastElement);
```
this is the right time to handle position of `ul` element so first image is visible and not the one you just cloned and prepended

please notice that when you load the website you may see rhinos image for a fraction of second (this is because `js` is not executed immediately), this is an unpleasant visual effect

lukaszfranczak replied:
updated in the newest commit

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,71 @@
+            },
+            {
+                complete: function () {
+                    if (visibleImageIndex > 6) {
```
`6` - magical number
your script should work with any number of images, don't hardcode this value
read it from `elementListLi`

this applies to all other hardcoded numerical values in this file: `404`, `2828` (let browser do the math!)

lukaszfranczak replied:
So obvious... Thanks for that!

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,71 @@
+    var elementListLi = $('li');
+    // console.log(elementListLi);
+
+    var elementUl = $('ul');
```
I noticed you prefixed above variables with `element`. Nothing wrong with it but more common way to say "hey, this is an element" is to start variable name with `$` (`$nextButton` for example).

lukaszfranczak replied:
thanks! I applied this rule to all other scripts

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,71 @@
+
+    elementNextButton.on('click', function () {
+        clearInterval(interval);
+        interval = setInterval(moveRight, 1000);
```
why `setInterval`? is it just for development purposes? I think clicking "Next" should change an image once and do nothing more.

lukaszfranczak replied:
Hm. I thought that if it's meant to be an infinite slider then it should work till it's stopped (in this case, due to available buttons, till it's switched to the opposite direction

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,71 @@
+                    }
+                }
+            });
+        visibleImageIndex--;
```
Please notice that in `moveRight` you updated `visibleImageIndex` right at the beginning of the function. I think this may cause the misbehavior you described in a PR comment.

lukaszfranczak replied:
That was my main problem - if the index is increased/decreased on beginning/end in both cases the script doesn't work well. At the moment the only problem is the gap. I'll take a look at it once more

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,71 @@
+            });
+    }
+
+    function moveLeft() {
```
Noticed `moveRight` and `moveLeft` functions are very similar? Whenever you find yourself writing the same code twice ask yourself seriously: can I express it once? Here it would be possible to create `move` function which accepts a parameter so `move(-1)` moves left and `move(1)` moves right (yes, `complete` would need an update then). Code deduplication is a way to simplify code, avoid bugs, to make code more maintainable.

lukaszfranczak replied:
Updated in the newest commit. Even though I included two if conditions in it so I'm not sure if it's the best solution

tomecko commented:
I've read code for `4 ProgressBar` and `6 Infinite Slider`. Let me know if further code is ready to be reviewed. You may also want to create separate PRs for each task and assign me as a reviewer when ready.

lukaszfranczak commented:
I'll take a look at the Inifinite slider tomorrow (I hope). One more thing I have a problem of is understanding the point 6 in parallax task. 

I see the issue with rectangles "jumping", but still the description of this point is so murky I don't understand it, and, due to that, I'm now sure how should I approach it

lukaszfranczak committed:
Modul 04 - infinite slider - slightly reworked

tomecko commented on 4_Zadania_dodatkowe/3.StickyMenu/js/app.js:
```diff
@@ -1,26 +1,23 @@
 $(document).ready(function () {
-   // console.log('test');
+    var $menu = $('.menu');
+    var $navPosition = $('nav').position().top;
```
here the variable holds numerical value, not an element
so prefixing variable name with `$` doesn't feel right

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+    var $prevButton = $('#prevPicture');
+    var $nextButton = $('#nextPicture');
+
+    var $liList = $('li');
```
it's not a "list"
more like `$sliderItem` or `$item`

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+    var $ul = $('ul');
+
+    var visibleImageIndex = 0;
+    var imageWidth = $('img').width()+4;   // I tried with outerwidth, css(width) etc but still I have to add 4 because of the space between images; reset of all elements also didn't help
```
see https://github.com/lukaszfranczak/jQueryTestRepo/pull/3#discussion_r188444341

lukaszfranczak replied:
You're a genius :) I was trying to find out what's wrong but found only the outerWidth() which didn't work. Now it's finally working :)

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+    $ul.append($firstLi);
+    $ul.prepend($lastLi);
+
+    $ul.css('right', imageWidth+'px').css('width', imageWidth*8+'px');
```
`8` - magical number, it should be calculated from `$liList` (or whatever you name it)

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+
+    $ul.css('right', imageWidth+'px').css('width', imageWidth*8+'px');
+
+    function move(parameter) {
```
`parameter` - not the best name, it conveys no meaning
maybe more like `step` or `change`

lukaszfranczak replied:
Changed to direction as this is the actual meaning :)

tomecko replied:
`direction` - I like it!

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+    function move(parameter) {
+        if (parameter===1) {
+            visibleImageIndex++;
+        }
```
maybe I'm missing something but I'd expect
`visibleImageIndex = visibleImageIndex + step`
instead of lines 21-23 AND 43-45
not sure why moving left case is handled at the end of the function

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+                    } else if (parameter===-1) {
+                        if (visibleImageIndex < 1) {
+                            visibleImageIndex = $liList.length;
+                            $ul.css('right', imageWidth*7);
```
again, `7` is a magical hardcoded value here

lukaszfranczak replied:
True. Changed to a $sliderItemsTotal.length (which equals to 8) minus 1 as I need to show the last picture and $sliderItemsTotal.length would move me further

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+            {
+                complete: function () {
+                    if (parameter===1) {
+                        if (visibleImageIndex > $liList.length) {
```
:+1: for using `$liList.length`

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+                            visibleImageIndex = 1;
+                            $ul.css('right', imageWidth);
+                        }
+                    } else if (parameter===-1) {
```
see line 29 and line 34
these `parameter` conditions are not needed I think
it's not possible that `visibleImageIndex > $liList.length` but `parameter` is `-1`

I think you what you should do here is to:
1. calculate new value of `visibleImageIndex` if it corresponds to cloned elements (and you got this logic right! lines 30-31 and lines 35-36)
2. if `visibleImageIndex` was reassigned then do `$ul.css('right', imageWidth * visibleImageIndex);`

tomecko commented on 4_Zadania_dodatkowe/6.InfiniteSlider/js/app.js:
```diff
@@ -0,0 +1,65 @@
+        clearInterval(interval);
+        interval = setInterval(function() {
+            move(1);
+            }, 1000);
```
I think
1. `infinite` means you can always click `right` (if you reach an end first image is displayed; you can click next button infinitely)
2. it does not mean slider is auto-playing

lukaszfranczak replied:
And I think that I'll ask the guys who prepared it as at the moment I don't know:)

tomecko replied:
I'm pretty sure I'm right though ;)

tomecko commented on 4_Zadania_dodatkowe/8.Parallax/js/app.js:
```diff
@@ -0,0 +1,46 @@
+        })
+        .each(function() {
+            $(this).css('top', $(this).data('y'))
+        });
```
you don't have to write `each` method multiple times

you can do:
```
.each(function() {
    $(this)
        .css('z-index', $(this).data('z'))
        .css('left', $(this).data('x'))
        .css('top', $(this).data('y'));
})
```
or even
```
.each(function() {
    $(this).css({
        'z-index': ...,
        'left': ...,
        'top': ...,
    });
})
```

lukaszfranczak replied:
Neat - like it :)

tomecko commented on 4_Zadania_dodatkowe/8.Parallax/js/app.js:
```diff
@@ -0,0 +1,46 @@
+            .each(function() {
+                $(this).css('left', $(this).data('x')+mouseMoveX*$(this).data('speed'))
+            })
+            .each(function() {
```
can be done in one `each`, see above

tomecko commented on 4_Zadania_dodatkowe/8.Parallax/js/app.js:
```diff
@@ -0,0 +1,46 @@
+        oldMousePositionY = event.offsetY;
+    });
+
+    $('.scene').on('mousemove', function (event) {
```
regarding `Punkt 6`:
1. notice `event` parameter in line 28
2. `even.target` should be an element that a cursor points at at the moment
3. they want you to check if `even.target` is one of `$('.element')`
4. if it's true then you're supposed to change `currentMousePositionX` and `currentMousePositionY` (add `left` and `top` css values from `even.target`)
5. I think it should be done before line 32

I didn't play with it, didn't test the code…

lukaszfranczak committed:
Parallax - finally working

Still - the problem with the rectangles (and the solution) is a bit unclear to me
